File: java/src/test/java/theatricalplays/StatementPrinterTests.java
package theatricalplays;

import org.junit.jupiter.api.Test;
import theatricalplays.play.Comedy;
import theatricalplays.play.Tragedy;

import java.util.List;

import static org.approvaltests.Approvals.verify;

class StatementPrinterTests {

    @Test
    void exampleStatement() {
        Invoice invoice = new Invoice("BigCo", List.of(
                new Performance(new Tragedy("Hamlet"), 55),
                new Performance(new Comedy("As You Like It"), 35),
                new Performance(new Tragedy("Othello"), 40)));

		PlayStatementCreator statementCreator = new PlayStatementCreator(new PlaintextStatementPrinter());
        var result = statementCreator.createStatement(invoice);

        verify(result);
    }

    //@Test
    //void statementWithNewPlayTypes() {
    //    Map<String, Play> plays = Map.of(
    //            "henry-v",  new Play("Henry V", PlayType.HISTORY),
    //            "as-like", new Play("As You Like It", PlayType.PASTORAL));

    //    Invoice invoice = new Invoice("BigCo", List.of(
    //            new Performance("henry-v", 53),
    //            new Performance("as-like", 55)));

    //    StatementPrinter statementPrinter = new StatementPrinter();
    //    Assertions.assertThrows(Error.class, () -> statementPrinter.print(invoice, plays));
    //}
}

File: csharp/TheatricalPlayersRefactoringKata.Tests/StatementPrinterTests.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using VerifyXunit;
using Xunit;

namespace TheatricalPlayersRefactoringKata.Tests
{
    public class StatementPrinterTests
    {
        [Fact]
        public Task test_statement_example()
        {
            var plays = new Dictionary<string, Play>();
            plays.Add("hamlet", new TragedyPlay("Hamlet"));
            plays.Add("as-like", new ComedyPlay("As You Like It"));
            plays.Add("othello", new TragedyPlay("Othello"));

            Invoice invoice = new Invoice("BigCo", new List<Performance>{new Performance("hamlet", 55),
                new Performance("as-like", 35),
                new Performance("othello", 40)});
            
            StatementPrinter statementPrinter = new StatementPrinter();
            var result = statementPrinter.Print(invoice, plays);

            return Verifier.Verify(result);
        }
        [Fact]
        public void test_statement_with_new_play_types()
        {
            // This test verifies that we can create new play types by extending the class hierarchy
            var plays = new Dictionary<string, Play>();
            plays.Add("henry-v", new TragedyPlay("Henry V"));
            plays.Add("as-like", new ComedyPlay("As You Like It"));

            Invoice invoice = new Invoice("BigCoII", new List<Performance>{new Performance("henry-v", 53),
                new Performance("as-like", 55)});
            
            StatementPrinter statementPrinter = new StatementPrinter();

            // This should work fine now since we're using concrete play types
            var result = statementPrinter.Print(invoice, plays);
            Assert.NotNull(result);
            Assert.Contains("BigCoII", result);
        }
    }
}
